Validation/RecoTrack/plugins/TrackFromSeedProducer.cc

```diff
@@ -0,0 +1,250 @@
+//
+
+//
+// constructors and destructor
+//
+TrackFromSeedProducer::TrackFromSeedProducer(const edm::ParameterSet& iConfig)
+{
+  //register your products
+  produces<reco::TrackCollection>();
+  produces<TrackingRecHitCollection>();
+  produces<reco::TrackExtraCollection>();
+  produces<std::vector<int> >();
+
+  // read parametes
+  edm::InputTag seedsTag(iConfig.getParameter<edm::InputTag>("src"));
+  edm::InputTag beamSpotTag(iConfig.getParameter<edm::InputTag>("beamSpot"));
+  tTRHBuilderName = iConfig.getParameter<std::string>("TTRHBuilder");
+  
+  //consumes
+  seedsToken = consumes<std::vector<TrajectorySeed> >(seedsTag);
+  beamSpotToken = consumes<reco::BeamSpot>(beamSpotTag);
+}
+
+
+TrackFromSeedProducer::~TrackFromSeedProducer()
+{
+ 
+   // do anything here that needs to be done at desctruction time
+   // (e.g. close files, deallocate resources etc.)
+
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void
+TrackFromSeedProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+   using namespace edm;
+   using namespace reco;
+   using namespace std;
+   
+   // output collection
+   auto_ptr<TrackCollection> tracks(new TrackCollection);
+   auto_ptr<TrackingRecHitCollection> rechits(new TrackingRecHitCollection);
+   auto_ptr<TrackExtraCollection> trackextras(new TrackExtraCollection);
+   auto_ptr<vector<int> > seedToTrack(new vector<int>());
+   
+   // product references 
+   TrackExtraRefProd ref_trackextras = const_cast<Event&>( iEvent ).getRefBeforePut<TrackExtraCollection>();
+   TrackingRecHitRefProd ref_rechits = const_cast<Event&>( iEvent ).getRefBeforePut<TrackingRecHitCollection>();
+
+   // input collection
+   Handle<vector<TrajectorySeed> > seeds;
+   iEvent.getByToken(seedsToken,seeds);
+
+   // beam spot
+   edm::Handle<reco::BeamSpot> beamSpot;
+   iEvent.getByToken(beamSpotToken,beamSpot);
+
+   // some objects to build to tracks
+   TSCBLBuilderNoMaterial tscblBuilder;
+
+   edm::ESHandle<TransientTrackingRecHitBuilder> tTRHBuilder;
+   iSetup.get<TransientRecHitRecord>().get(tTRHBuilderName,tTRHBuilder);
+
+   edm::ESHandle<MagneticField> theMF;
+   iSetup.get<IdealMagneticFieldRecord>().get(theMF);
+
+   // create tracks from seeds
+   int nfailed  = 0;
+   for (auto const & seed : *seeds.product()){
+     // try to create a track
+     TransientTrackingRecHit::RecHitPointer lastRecHit = tTRHBuilder->build(&*(seed.recHits().second-1));
+     TrajectoryStateOnSurface state = trajectoryStateTransform::transientState( seed.startingState(), lastRecHit->surface(), theMF.product());
+     TrajectoryStateClosestToBeamLine tsAtClosestApproachSeed = tscblBuilder(*state.freeState(),*beamSpot);//as in TrackProducerAlgorithm
+     if(!(tsAtClosestApproachSeed.isValid())){
+       edm::LogVerbatim("SeedValidator")<<"TrajectoryStateClosestToBeamLine not valid";
+       seedToTrack->push_back(-1);
+       nfailed++;
+       continue;
```

Comment:
That's true, for fake rate plots the track properties are used.
However, I am not too much interested in the fake rate of seeds at this
moment.
I really want to learn about the matched tracks.
I'm not sure what the interests of others are...


On Mon, Jun 29, 2015 at 11:20 AM, Matti Kortelainen <
notifications@github.com> wrote:

> In Validation/RecoTrack/plugins/TrackFromSeedProducer.cc
> <https://github.com/cms-sw/cmssw/pull/9917#discussion_r33446648>:
>
> > +
> > +   edm::ESHandle<MagneticField> theMF;
> > +   iSetup.get<IdealMagneticFieldRecord>().get(theMF);
> > +
> > +   // create tracks from seeds
> > +   int nfailed  = 0;
> > +   for (auto const & seed : *seeds.product()){
> > +     // try to create a track
> > +     TransientTrackingRecHit::RecHitPointer lastRecHit = tTRHBuilder->build(&*(seed.recHits().second-1));
> > +     TrajectoryStateOnSurface state = trajectoryStateTransform::transientState( seed.startingState(), lastRecHit->surface(), theMF.product());
> > +     TrajectoryStateClosestToBeamLine tsAtClosestApproachSeed = tscblBuilder(*state.freeState(),*beamSpot);//as in TrackProducerAlgorithm
> > +     if(!(tsAtClosestApproachSeed.isValid())){
> > +       edm::LogVerbatim("SeedValidator")<<"TrajectoryStateClosestToBeamLine not valid";
> > +       seedToTrack->push_back(-1);
> > +       nfailed++;
> > +       continue;
>
> Hmm, actually, I think, the track is used as a bit more than just a rechit
> container (e.g. fake rate vs. pT), for which the fit is needed. I don't
> really know what to do to avoid the failures (code is the same as currently
> in TrackerSeedValidator), maybe @rovere <https://github.com/rovere> or
> @VinInn <https://github.com/VinInn> would have some ideas (now that
> they're in the loop anyway)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9917/files#r33446648>.
>